src/emit/pretty_printer/trivia.cpp

```diff
@@ -3,67 +3,101 @@
+/// Comments don't get a trailing hardline
+/// Breaks are reduced by 1 (since the block starts with \n).
+auto formatLastTrailing(const ast::Trivia &t) -> Doc
+{
+    return std::visit(
+      common::Overload{ [](const ast::Comment &c) -> Doc { return Doc::text(c.text); },
+                        [](const ast::Break &p) -> Doc {
+                            return Doc::hardlines(p.blank_lines > 0 ? p.blank_lines - 1 : 0);
+                        } },
+      t);
 }
 
-auto printTrailingTriviaList(const std::span<const ast::Trivia> list) -> Doc
+/// @brief Generic builder for a block of trivia.
+/// @tparam Formatter Function to handle the very last item in the list.
+template<typename Formatter>
+auto buildTrivia(std::span<const ast::Trivia> trivia,
+                 bool suppress,
+                 Doc prefix,
+                 Formatter format_last) -> Doc
 {
-    if (list.empty()) {
+    auto view = trivia | std::views::filter(shouldKeep(suppress));
+
+    auto it = view.begin();
+    if (it == view.end()) {
         return Doc::empty();
     }
 
-    const Doc result = std::ranges::fold_left(
-      list | std::views::take(list.size() - 1) | std::views::transform(printTrivia),
-      Doc::line(),
-      std::plus<>());
+    // Start with the specific prefix (Empty for Leading, Hardline for Trailing)
+    Doc doc = std::move(prefix);
 
-    auto com = [](const ast::Comment &c) -> Doc {
-        // `hardlines(0)` makes it so it can't be flattened
-        return Doc::text(c.text) + Doc::hardlines(0);
-    };
-    auto par = [](const ast::ParagraphBreak &p) -> Doc {
-        return Doc::hardlines(std::max(p.blank_lines - 1, 0U));
-    };
+    auto pending = *it++;
 
-    return result + std::visit(common::Overload{ com, par }, list.back());
+    // Process all items except the last one
+    for (const auto &item : std::ranges::subrange(it, view.end())) {
+        doc += formatTrivia(std::exchange(pending, item));
+    }
+
+    // Apply the specific strategy for the last item
+    return doc + format_last(pending);
 }
 
 } // namespace
 
-auto PrettyPrinter::withTrivia(const ast::NodeBase &node, Doc core_doc) -> Doc
+auto PrettyPrinter::withTrivia(const ast::NodeBase &node, Doc core, bool suppress) -> Doc
 {
-    if (!node.trivia) {
-        return core_doc;
+    if (!node.hasTrivia()) {
+        return core;
     }
 
-    const auto &trivia = *node.trivia;
+    Doc result = Doc::empty();
 
-    const Doc leading = std::ranges::fold_left(
-      trivia.leading | std::views::transform(printTrivia), Doc::empty(), std::plus<>());
+    // 1. Leading Trivia
+    result += buildTrivia(node.getLeading(), suppress, Doc::empty(), formatTrivia);
 
-    const Doc inline_comment
-      = trivia.inline_comment
-        ? Doc::text(" ") + Doc::text(trivia.inline_comment->text) + Doc::hardlines(0)
-        : Doc::empty();
+    // 2. Core Doc
+    result += core;
+
+    // 3. Inline Comment
+    if (auto txt = node.getInlineComment()) {
```

Comment:
```suggestion
    if (auto comment = node.getInlineComment()) {
```

If `txt` refers to `text` why not just write `text`.  IMO  `text` is not really meaningful either for a comment

Reply:
-> comment